README.md
# IPK Project 2 - ZETA: Network sniffer

        Author:  Maksim Kalutski (xkalut00)
        Date:    2023-04-17

This is a C++ program that implements a network analyzer that captures and filters packets on a specific network
interface. The program is able to display the following protocols: TCP, UDP, ARP, ICMPv4, ICMPv6, NDP, IGMP, and MLD. Packets
can be filtered by port number, protocol and interface. You can also specify the number of packets to display.

## Written in

* [C++20 language](https://en.wikipedia.org/wiki/C%2B%2B)

## Theory

### Transmission Control Protocol (TCP)
A TCP segment consists of a segment header and a data section. The segment header contains 10 mandatory fields, and 
an optional extension field (Options). The TCP header can range in size from 20 to 60 bytes, depending on the size 
of the options field. Some of the fields in the TCP header include source port number, destination port number, 
sequence number, acknowledgement number and checksum.

                                  0                   1                   2                   3
                                  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |          Source Port          |       Destination Port        |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |                        Sequence Number                        |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |                    Acknowledgment Number                      |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |  Data |       |C|E|U|A|P|R|S|F|                               |
                                 | Offset| Rsrvd |W|C|R|C|S|S|Y|I|            Window             |
                                 |       |       |R|E|G|K|H|T|N|N|                               |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |           Checksum            |         Urgent Pointer        |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |                           [Options]                           |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                 |                                                               :
                                 :                             Data                              :
                                 :                                                               |
                                 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


### User Datagram Protocol (UDP)
UDP segment consists of an 8-byte header and variable length data. The first four bytes of the UDP header store the
port numbers for the source and destination. The next two bytes store the length (in bytes) of the segment (including 
the header). The final two bytes of the UDP header is the checksum, a field that’s used by the sender and receiver t
o check for data corruption.

                                             0      7 8     15 16    23 24    31
                                            +--------+--------+--------+--------+
                                            |     Source      |   Destination   |
                                            |      Port       |      Port       |
                                            +--------+--------+--------+--------+
                                            |                 |                 |
                                            |     Length      |    Checksum     |
                                            +--------+--------+--------+--------+
                                            |
                                            |          data octets ...
                                            +---------------- ...

### Address Resolution Protocol (ARP)
ARP is a communication protocol used for discovering the link layer address, such as a MAC address, associated with a 
given internet layer address, typically an IPv4 address. The ARP packet size is 28 bytes and contains fields such as 
Hardware type (HTYPE), Protocol type (PTYPE), Hardware address length (HLEN), Protocol address length (PLEN), Operation 
(OPER), Sender hardware address (SHA), Sender protocol address (SPA), Target hardware address (THA) and Target protocol 
address (TPA).
                              
                                            0        7        15       23       31
                                            +--------+--------+--------+--------+
                                            |       HT        |        PT       |
                                            +--------+--------+--------+--------+
                                            |  HAL   |  PAL   |        OP       |
                                            +--------+--------+--------+--------+
                                            |         S_HA (bytes 0-3)          |
                                            +--------+--------+--------+--------+
                                            | S_HA (bytes 4-5)|S_L32 (bytes 0-1)|
                                            +--------+--------+--------+--------+
                                            |S_L32 (bytes 2-3)|S_NID (bytes 0-1)|
                                            +--------+--------+--------+--------+
                                            |         S_NID (bytes 2-5)         |
                                            +--------+--------+--------+--------+
                                            |S_NID (bytes 6-7)| T_HA (bytes 0-1)|
                                            +--------+--------+--------+--------+
                                            |         T_HA (bytes 3-5)          |
                                            +--------+--------+--------+--------+
                                            |         T_L32 (bytes 0-3)         |
                                            +--------+--------+--------+--------+
                                            |         T_NID (bytes 0-3)         |
                                            +--------+--------+--------+--------+
                                            |         T_NID (bytes 4-7)         |
                                            +--------+--------+--------+--------+

### Internet Control Message Protocol
ICMP is a supporting protocol in the Internet protocol suite used by network devices to send error messages and 
operational information indicating success or failure when communicating with another IP address. All ICMP packets have 
an 8-byte header and variable-sized data section. The first 4 bytes of the header have fixed format, while the last 
4 bytes depend on the type/code of that ICMP packet.

                               0                   1                   2                   3
                               0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |     Type      |     Code      |           unused              |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |      Internet Header + 64 bits of Original Data Datagram      |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

### Internet Group Management Protocol
IGMP is a protocol that allows several devices to share one IP address so they can all receive the same data.
IGMP is a network layer protocol used to set up multicasting on networks that use the Internet Protocol version 4 
(IPv4). IGMP messages have an 8-byte header and variable-sized data section.

                               0                   1                   2                   3
                               0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |  Type = 0x11  | Max Resp Code |           Checksum            |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |                         Group Address                         |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              | Resv  |S| QRV |     QQIC      |     Number of Sources (N)     |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |                       Source Address [1]                      |
                              +-                                                             -+
                              |                       Source Address [2]                      |
                              +-                              .                              -+
                              .                               .                               .
                              .                               .                               .
                              +-                                                             -+
                              |                       Source Address [N]                      |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


### Internet Control Message Protocol (ICMPv6)
ICMPv6 is an integral part of IPv6 and performs error reporting and diagnostic functions. ICMPv6 has a framework for 
extensions to implement new features. Several extensions have been published, defining new ICMPv6 message types as well 
as new options for existing ICMPv6 message types. Neighbor Discovery Protocol (NDP) is a node discovery protocol based 
on ICMPv6 which replaces and enhances functions of ARP1. Multicast Listener Discovery (MLD) is used by IPv6 routers 
for discovering multicast listeners on a directly attached link, much like Internet Group Management Protocol (IGMP) 
is used in IPv41.

                               0                   1                   2                   3
                               0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |     Type      |     Code      |          Checksum             |
                              +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                              |                                                               |
                              +                         Message Body                          +
                              |                                                               |


## Prerequisites
Before using the program, you need to have **g++**, **make** and **libpcap** installed on your **UNIX** operating system.

If you are using **Ubuntu** system you can install `g++` and `make` by running the following command:
```console
$ sudo apt-get install build-essential
```

To install `libpcap` on your system, run the following command:
```console
$ sudo apt install libpcap-dev
```


## Build
To build the program, run the following command:
```console
$ make
```

## Usage
```console
$ sudo ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}
```

### Command-line arguments
| Argument                  | Description                                                                                                           |
|---------------------------|-----------------------------------------------------------------------------------------------------------------------|
| `-i` <br /> `--interface` | Set interface device to sniff from. <br /> If this parameter is not specified a list of active interfaces is printed. |
| `-p port`                 | Filter packets on the given interface by port.                                                                        |
| `-n num`                  | The number of packets to display.                                                                                     |
| `-t` <br /> `--tcp`       | Display TCP packets.                                                                                                  |
| `-u` <br /> `--udp`       | Display UDP packets.                                                                                                  |
| `--icmp4 `                | Display only ICMPv4 packets.                                                                                          |
| `--icmp6`                 | Display only ICMPv6 echo request/response.                                                                            |
| `--arp`                   | Display only ARP frames.                                                                                              |
| `--ndp`                   | Display only ICMPv6 NDP packets.                                                                                      |
| `--igmp`                  | Display only IGMP packets.                                                                                            |
| `--mld`                   | Display only MLD packets.                                                                                             |


## Usage examples
```console
$ ./ipk-sniffer
```

```console
$ sudo ./ipk-sniffer -i eth0 
```

```console
$ sudo ./ipk-sniffer -i eth0 -p 23 --tcp -n 2
```

```console
$ sudo ./ipk-sniffer -i eth0 -p 22 --tcp --udp --icmp4 --icmp6 --arp --ndp --igmp --mld 
```


## Output examples

```
./ipk-sniffer
./ipk-sniffer -i

eth0
any
lo
dummy0
tunl0
sit0
bluetooth-monitor
nflog
nfqueue
dbus-system
dbus-session
bond0

```

```
$ sudo ./ipk-sniffer -i eth0
timestamp: 2021-03-19T18:42:52.362+01:00
src MAC: 00:1c:2e:92:03:80
dst MAC: 00:1b:3f:56:8a:00
frame length: 512 bytes
src IP: 147.229.13.223
dst IP: 10.10.10.56
src port: 4093
dst port: 80

0x0000: 00 19 d1 f7 be e5 00 04 96 1d 34 20 08 00 45 00 ........ ..4 ..
0x0010: 05 a0 52 5b 40 00 36 06 5b db d9 43 16 8c 93 e5 ..R[@.6. [..C....
0x0020: 0d 6d 00 50 0d fb 3d cd 0a ed 41 d1 a4 ff 50 18 .m.P..=. ..A...P.
0x0030: 19 20 c7 cd 00 00 99 17 f1 60 7a bc 1f 97 2e b7 . ...... .`z.....
0x0040: a1 18 f4 0b 5a ff 5f ac 07 71 a8 ac 54 67 3b 39 ....Z._. .q..Tg;9
0x0050: 4e 31 c5 5c 5f b5 37 ed bd 66 ee ea b1 2b 0c 26 N1.\_.7. .f...+.&
0x0060: 98 9d b8 c8 00 80 0c 57 61 87 b0 cd 08 80 00 a1 .......W a.......

```

## Libraries
The program uses the following libraries:

- `iostream`: for input and output operations.
- `cstdlib`: for functions such as strtol and strtoul.
- `cstring`: for functions such as strcmp.
- `iomanip`: for manipulating input and output formats.
- `string`: for string handling functions.
- `ctime`: for date and time functions.
- `cctype`: for character classification functions.
- `csignal`: for signal handling.


- `pcap.h`: for packet capture functions.


- `netinet/tcp.h`: for TCP header structure.
- `netinet/udp.h`: for UDP header structure.
- `netinet/in.h`: for internet address structure.
- `netinet/ip.h`: for IP header structure.

## UML Diagram

                                       +---------------------------------+
                                       |           ipk-sniffer           |
                                       +---------------------------------+
                                       | - handle: pcap_t*               |
                                       +---------------------------------+
                                       | - EXIT_SUC: int = 0             |
                                       | - EXIT_ARGS: int = 1            |
                                       | - EXIT_DEVS: int = 2            |
                                       | - EXIT_PCAP: int = 3            |
                                       | - EXIT_FILT: int = 4            |
                                       | - EXIT_PAC: int = 5             |
                                       | - EXIT_SIG: int = 6             |
                                       +---------------------------------+
                                       | - args: struct                  |
                                       |   - interface: char*            |
                                       |   - numPackets: size_t          |
                                       |   - port: int                   |
                                       |   - tcp: int                    |
                                       |   - udp: int                    |
                                       |   - icmp4: int                  |

CHANGELOG.md
# Changelog

## Version 0.1.0 (2023-4-9)

### Minor changes
*   init: [`c1536c5520`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/c1536c552087b85d6a821afff9c1ddaf5fe0d746)

## Version 0.1.1 (2023-4-9)

### Major changes
*   added makefile: [`c7a4b63944`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/c7a4b63944fb4881a183020c5aa72e04064f9cc5)

### Minor changes
*   added readme and changelog files: [`947e408217`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/947e408217a38038588b19c020d36c7602bf7475)

## Version 0.2.0 (2023-4-9)

### Major changes
*   added functions to parse arguments and print usage: [`ea52b9c4fd`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/ea52b9c4fdeb4af739c63e31fd46e54fcf48c3d9)

## Version 0.2.1 (2023-4-14)

### Major changes
*   full make over of parse arguments function: [`36640d60b0`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/36640d60b05af0ea5bc13639e43996ee22f85777)

## Version 0.3.0 (2023-4-14)

### Major changes
*   added function for showing active interfaces: [`aedb21fbbd`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/aedb21fbbdfdade421ee5f27823d1febfa81385b)

### Minor changes
*   added support of the pcap to the Makefile: [`3079e9d79e`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/3079e9d79e88fdb05fbdcfd975602f9cc4e1e828)

## Version 0.4.0 (2023-4-15)

### Major changes
*   added function for creating network packet filter: [`5e369662f1`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/5e369662f1ecfa43cd6df543c4b63292f8dbf67d)

## Version 0.4.1 (2023-4-16)

### Major changes
*   full make over of the function for creating network packet filter: [`414ceb4d6e`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/414ceb4d6e0b8bda7d0586335b1a9483c8cd1833)

## Version 0.5.0 (2023-4-16)

### Major changes
*   added function for unction for filtering and printing packets: [`dacb2e5fce`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/dacb2e5fcee5d115e75531fb94e0ef6ddcf6b272)

## Version 0.6.0 (2023-4-16)

### Major changes
*   added function for ctrl-c signal handler: [`d0894155ec`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/d0894155ececf20f272d1709a8677cbab7f85532)

### Minor changes
*   added exit codes for a program: [`72fb7a143e`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/72fb7a143ebc08c202285180d060016cf96bd9f4)

## Version 0.7.0 (2023-4-17)

### Major changes
*   added summary, usage, libraries, uml diagram, functions descriptions and bibliography to README.md file: [`3c511dc5d8`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/3c511dc5d85a65f46e8174ebc941981bfb94866e)
*   added changes to CHANGELOG.md file: [`232f1321e1`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/232f1321e177673986b916d018b3ccbd37aa2ccb)

### Bug fixes
*   changed milliseconds to seconds in timestamp to match format: [`2bd5a276fc`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/2bd5a276fc104efb3407d66fb4315bbceb2fae62)
*   removed the extra space when printing the packet: [`23493a9901`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/23493a9901716eb1025c73db673b97a8b8279e92)

## Version 0.8.0 (2023-4-17)

### Major changes
*   added testing and theory sections to the README.md file: [`c326f9c703`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/df95592464c51843b80f82b228c600a1d9c6c24f)
*   added tests folder with python scripts and testing screenshots: [`d9ae21d2fe`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/d9ae21d2fefa1ed1722be874c028d5423cfbf5dd)

### Minor changes
*   added header and comments for each function: [`df95592464`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/df95592464c51843b80f82b228c600a1d9c6c24f)

## Version 0.8.1 (2023-4-17)

### Minor changes
*   adds tests for empty output to README.md file: [`cf9d813b4c`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/cf9d813b4cc4422965b82090b38423f0361df9fb)

### Bug fixes
*   added missing if condition when only mld and ndp selected: [`3ec105bc56`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/3ec105bc569679d47ef3cdea0dc7bfc43fd17597)
*   removed the additional filter print used for testing [`c25dc27404`](https://git.fit.vutbr.cz/xkalut00/ipk-project-2/commit/c25dc27404fd9f2eacc2f0d6cec95ca67dada1ca)

## Version 1.0.0 (2023-4-17)
*   This is the release version.




ipk-sniffer.cpp
/**
 * IPK Project 2 - ZETA: Network sniffer
 * @author
 *   xkalut00, Maksim Kalutski
 *
 * @file ipk-sniffer.cpp
 * @brief A network analyzer that captures and filters packets on a specific network interface.
 * @date 17.04.2023
 */

#include <iostream>
#include <cstdlib>
#include <cstring>
#include <iomanip>
#include <string>
#include <ctime>
#include <cctype>
#include <csignal>

#include <pcap.h>

#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/in.h>
#include <netinet/ip.h>

using namespace std;

pcap_t *handle;

#define EXIT_SUC 0
#define EXIT_ARGS 1
#define EXIT_DEVS 2
#define EXIT_PCAP 3
#define EXIT_FILT 4
#define EXIT_PAC 5
#define EXIT_SIG 6

struct args {
    char *interface;
    size_t numPackets;
    int port;
    int tcp;
    int udp;
    int icmp4;
    int icmp6;
    int arp;
    int ndp;
    int igmp;
    int mld;
};

/**
 * Prints usage of the program.
 */
void printUsage() {
    cout << "Usage:\n"
            "   ./ipk-sniffer [-i interface | --interface interface] {-p port [--tcp|-t] [--udp|-u]} [--arp] [--icmp4] [--icmp6] [--igmp] [--mld] {-n num}\n"
            "\n"
            "Options:\n"
            "   -i|--interface          Set interface device to sniff from.\n"
            "                           If this parameter is not specified a list of active interfaces is printed.\n"
            "   -p port                 Filter packets on the given interface by port.\n"
            "   -n num                  The number of packets to display.\n"
            "   -t|--tcp                Display TCP packets.\n"
            "   -u|--udp                Display UDP packets.\n"
            "   --icmp4                 Display only ICMPv4 packets.\n"
            "   --icmp6                 Display only ICMPv6 echo request/response.\n"
            "   --arp                   Display only ARP frames.\n"
            "   --ndp                   Display only ICMPv6 NDP packets.\n"
            "   --igmp                  Display only IGMP packets.\n"
            "   --mld                   Display only MLD packets." << endl;
}


/**
 * Parses arguments from command line
 * @param args struct to store parsed arguments
 * @param argc number of arguments
 * @param argc array of arguments
 */
bool parseArgs(struct args *args, int argc, char *argv[]) {
    if (argc == 1) {
        return true;
    }

    if (argc == 2) {
        if (strcmp("-i", argv[1]) == 0) {
            return true;
        } else {
            cerr << "ERROR:Invalid argument: " << argv[1] << "." << endl;
            return false;
        }
    }

    for (int i = 1; i < argc; ++i) {
        char *arg = nullptr;
        if (i + 1 < argc) {
            arg = argv[i + 1];
        }

        if (strcmp("-i", argv[i]) == 0 || strcmp("--interface", argv[i]) == 0) {
            if (!arg) {
                cerr << "ERROR:Option " << argv[i] << " requires an argument." << endl;
                return false;
            }
            args->interface = arg;
            ++i;

        } else if (strcmp("-p", argv[i]) == 0) {
            if (!arg) {
                cerr << "ERROR:Option " << argv[i] << " requires an argument." << endl;
                return false;
            }
            char *rest;
            args->port = strtol(arg, &rest, 10);
            if (*rest != '\0' || args->port < 0 || args->port > 65535) {
                cerr << "ERROR:Invalid port number." << endl;
                return false;
            }
            ++i;

        } else if (strcmp("--tcp", argv[i]) == 0) {
            args->tcp = 1;

        } else if (strcmp("--udp", argv[i]) == 0) {
            args->udp = 1;

        } else if (strcmp("--arp", argv[i]) == 0) {
            args->arp = 1;

        } else if (strcmp("--icmp4", argv[i]) == 0) {
            args->icmp4 = 1;

        } else if (strcmp("--icmp6", argv[i]) == 0) {
            args->icmp6 = 1;

        } else if (strcmp("--ndp", argv[i]) == 0) {
            args->ndp = 1;

        } else if (strcmp("--igmp", argv[i]) == 0) {
            args->igmp = 1;

        } else if (strcmp("--mld", argv[i]) == 0) {
            args->mld = 1;

        } else if (strcmp("-n", argv[i]) == 0) {
            if (!arg) {
                cerr << "ERROR:Option " << argv[i] << " requires an argument."<< endl;
                return false;
            }
            char *rest;
            args->numPackets = strtoul(arg, &rest, 10);
            if (*rest != '\0' || args->numPackets <= 0) {
                cerr << "ERROR:Number of packets must be a positive integer." << endl;
                return false;
            }
            ++i;

        } else {
            cerr << "ERROR:Invalid argument: " << argv[i] << "." << endl;
            return false;
        }
    }

    return true;
}

/**
 * Prints list of active interfaces.
 * @param errbuf buffer for error messages
 */
void printActiveInterfaces(char *errbuf) {
    pcap_if_t *allInterfaces;
    if (pcap_findalldevs(&allInterfaces, errbuf) == PCAP_ERROR) {
        cerr << "ERROR:Could not get list of interfaces: " << errbuf << endl;
        exit(EXIT_DEVS);
    }
    if (!allInterfaces) {
        cerr << "ERROR:No active interfaces found." << endl;
        exit(EXIT_DEVS);
    }
    cout << endl;
    for (pcap_if_t *interface = allInterfaces; interface; interface = interface->next) {
        cout << interface->name << endl;
    }
    cout << endl;
    pcap_freealldevs(allInterfaces);
    exit(EXIT_SUC);
}

/**
 * Creates filter string from parsed arguments.
 * @param args struct with parsed arguments
 * @return filter string
 */
string createFilter(struct args *args) {
    string filter;
    string port = to_string(args->port);
    bool orSwitch = false;

    auto appendFilter = [&](const string& s) {
        if (orSwitch) {
            filter.append(" or ");
        }
        filter.append(s);
        orSwitch = true;
    };

    if (args->port >= 0) {
        if (args->udp) appendFilter("(udp and port " + port + ")");
        if (args->tcp) appendFilter("(tcp and port " + port + ")");
    } else {
        if (args->udp) appendFilter("udp");
        if (args->tcp) appendFilter("tcp");
    }

    if (args->arp) appendFilter("arp");
    if (args->icmp4) appendFilter("icmp");
    if (args->igmp) appendFilter("igmp");

    if (args->icmp6) {
        if (args->ndp && args->mld) {
            appendFilter("icmp6");
        } else if (args->ndp && !args->mld) {
            appendFilter("((icmp6 and (icmp6[0] == 128 or icmp6[0] == 129)) or (icmp6 and (icmp6[0] >= 133 and icmp6[0] <= 137)))");
        } else if (!args->ndp && args->mld) {
            appendFilter("((icmp6 and (icmp6[0] == 128 or icmp6[0] == 129)) or (icmp6 and (icmp6[0] >= 130 and icmp6[0] <= 132)))");
        } else {
            appendFilter("(icmp6 and (icmp6[0] == 128 or icmp6[0] == 129))");
        }
    }

    if (args->ndp && args->mld) {
        if (!args->icmp6) {
            appendFilter("(icmp6 and (icmp6[0] >= 130 and icmp6[0] <= 137))");
        }
    }

    if (args->ndp) {
        if (!args->icmp6 && !args->mld) {
            appendFilter("(icmp6 and (icmp6[0] >= 133 and icmp6[0] <= 137))");
        }
    }

    if (args->mld) {
        if (!args->icmp6 && !args->ndp) {
            appendFilter("(icmp6 and (icmp6[0] >= 130 and icmp6[0] <= 132))");
        }
    }

    if (!orSwitch) {
        if (args->port >= 0) {
            filter.append("(udp and port " + port + ") or (tcp and port " + port + ")");
        } else {
            filter.append("tcp or udp");
        }
        filter.append(" or arp or icmp or igmp or icmp6");
    }

    return filter;
}

/**
 * Helper function for the printPacket() function. It is used to parse the time zone offset.
 * @param timestamp timestamp string
 * @param microseconds microseconds
 */
string parseZoneOffset(const char* timestamp, int microseconds) {
    char buffer[40];
    sprintf(buffer, timestamp, microseconds);
    string str(buffer);
    str.insert(str.length() - 2, ":");
    return str;
}

/**
 * Prints and filter the contents of each packet captured by the sniffer.
 * @param header packet header
 * @param packetData packet data
 */
void printPacket(const struct pcap_pkthdr* header, const u_char* packetData) {
    time_t timestamp = header->ts.tv_sec;
    struct tm *localTime = localtime(&timestamp);
    char timestampStr[30];
    strftime(timestampStr, sizeof(timestampStr), "%Y-%m-%dT%H:%M:%S.%%03u%z", localTime);
    cout << "timestamp: " << parseZoneOffset(timestampStr, header->ts.tv_usec / 1000) << endl;

    cout << "src MAC: " << hex << setfill('0');
    for (int i = 0; i < 6; i++) {
        cout << setw(2) << (int)packetData[i];
        if (i < 5) {
            cout << ":";
        }
    }
    cout << dec << endl;

    cout << "dst MAC: " << hex << setfill('0');
    for (int i = 6; i < 12; i++) {
        cout << setw(2) << (int)packetData[i];
        if (i < 11) {
            cout << ":";
        }
    }
    cout << dec << endl;

    u_short etherType = (packetData[12] << 8) | packetData[13];
    const u_char *payload = packetData + 14;
    int payloadSize = header->len - 14;

    cout << "frame length: " << header->len << " bytes" << endl;

    if (etherType == 0x0800) {
        const struct ip *ipHeader = (struct ip *) payload;
        int ipHeaderSize = ipHeader->ip_hl * 4;

        char srcIP[INET6_ADDRSTRLEN], dstIP[INET6_ADDRSTRLEN];
        inet_ntop(AF_INET, &(ipHeader->ip_src), srcIP, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, &(ipHeader->ip_dst), dstIP, INET_ADDRSTRLEN);
        cout << "src IP: " << srcIP << "\n";
        cout << "dst IP: " << dstIP << "\n";

        u_short srcPort = 0, dstPort = 0;
        const u_char *transportPayload = payload + ipHeaderSize;
        if (ipHeader->ip_p == IPPROTO_TCP) {
            const struct tcphdr *tcpHeader = (struct tcphdr *) transportPayload;
            srcPort = ntohs(tcpHeader->th_sport);
            dstPort = ntohs(tcpHeader->th_dport);
        } else if (ipHeader->ip_p == IPPROTO_UDP) {
            const struct udphdr *udpHeader = (struct udphdr *) transportPayload;
            srcPort = ntohs(udpHeader->uh_sport);
            dstPort = ntohs(udpHeader->uh_dport);
        }
        if (srcPort || dstPort) {
            cout << "src port: " << srcPort << "\n";
            cout << "dst port: " << dstPort << "\n";
        }
    }

    cout << endl;

    for (int i = 0; i < header->len; i += 16) {
        cout << "0x" << hex << setw(4) << setfill('0') << i << ": ";

        for (int j = 0; j < 16; j++) {
            if (i + j < header->len) {
                cout << hex << setw(2) << setfill('0') << static_cast<int>(packetData[i + j]) << " ";
            } else {
                cout << "   ";
            }
        }

        for (int j = 0; j < 16; j++) {
            if (i + j < header->len) {
                if (isprint(packetData[i + j])) {
                    cout << packetData[i + j];
                } else {
                    cout << ".";
                }
            } else {
                cout << " ";
            }
            if ((j + 1) % 8 == 0) {
                cout << " ";
            }
        }

        cout << endl;
    }
    cout << endl;
}

/**
 * Function to catch Ctrl-c signal
 */
void signalHandler(int signal) {
    pcap_close(handle);
    cout << endl;
    exit(EXIT_SIG);
}

int main(int argc, char** argv) {
    struct args args = {0};
    args.port = -1;
    args.numPackets = 1;

    if (!parseArgs(&args, argc, argv)) {
        printUsage();
        exit(EXIT_ARGS);
    }

    char errbuf[PCAP_ERRBUF_SIZE];
    if (pcap_init(PCAP_CHAR_ENC_LOCAL, errbuf) == -1) {
        cerr << "ERROR:Could not initialize pcap: " << errbuf << endl;
        exit(EXIT_PCAP);
    }

    if(!args.interface) {
        printActiveInterfaces(errbuf);
    }

    handle = pcap_create(args.interface, errbuf);
    if (!handle) {
        cerr << "ERROR:Could not create interface " << args.interface << ": " << errbuf << endl;
        exit(EXIT_PCAP);
    }

    pcap_set_snaplen(handle, BUFSIZ);
    pcap_set_promisc(handle, 1);
    pcap_set_timeout(handle, 1000);

    if (pcap_activate(handle) < 0) {
        cerr << "ERROR:Could not activate interface " << args.interface << ": " << pcap_geterr(handle) << endl;
        exit(EXIT_PCAP);
    }

    struct bpf_program fp;
    bpf_u_int32 mask;
    bpf_u_int32 net;

    if (pcap_lookupnet(args.interface, &net, &mask, errbuf) == -1) {
        cerr << "ERROR:Could not get netmask for interface " << args.interface << ": " << errbuf << endl;
        exit(EXIT_PCAP);
    }

    if (pcap_datalink(handle) != DLT_EN10MB && pcap_datalink(handle) != DLT_LINUX_SLL) {
        cerr << "ERROR:Interface " << args.interface << " is not an Ethernet interface." << endl;
        exit(EXIT_PCAP);
    }

    string filters = createFilter(&args);

    signal(SIGINT, signalHandler);

    if (pcap_compile(handle, &fp, filters.c_str(), 0, net) == PCAP_ERROR) {
        cerr << "ERROR:Could not compile filter: " << pcap_geterr(handle) << endl;
        pcap_close(handle);
        exit(EXIT_FILT);
    }

    if (pcap_setfilter(handle, &fp) == PCAP_ERROR) {
        cerr << "ERROR:Could not set filter: " << pcap_geterr(handle) << endl;
        pcap_close(handle);
        exit(EXIT_FILT);
    }

    pcap_freecode(&fp);

    struct pcap_pkthdr *pktHeader = NULL;
    const u_char *pktData = NULL;
    int res = 0;
    for (int i = 0; i < args.numPackets; ++i) {
        res = pcap_next_ex(handle, &pktHeader, &pktData);
        if (res == 0) {
            cerr << "ERROR:Timeout while waiting for packet." << endl;
            pcap_close(handle);
            exit(EXIT_PAC);
        } else if (res == PCAP_ERROR) {
            cerr << "ERROR:Error while waiting for packet: " << pcap_geterr(handle) << endl;
            pcap_close(handle);
            exit(EXIT_PAC);
        } else if (res == PCAP_ERROR_BREAK) {
            cerr << "ERROR:Loop terminated by pcap_breakloop()." << endl;
            pcap_close(handle);
            exit(EXIT_PAC);
        }
        printPacket(pktHeader, pktData);
    }

    pcap_close(handle);
    exit(EXIT_SUC);
}
